### PR TITLE
DS-4418 remove uuid from resource policy rest

### PR DESCRIPTION
**Pull request by Micheleboychuk**
As by https://jira.lyrasis.org/browse/DS-4418
added minimal IT to the upload step to verify that the changes don't break the functionality as used by angular

### CODE REVIEW

**Comment by abollini** [comment]
please note that this PR is build on top of #2604 so it should be reviewed after than the previous one is merged.

The conflict is related to the changes at the master done in 
https://github.com/DSpace/DSpace/commit/f8ab7def82783da2325c55fbf88d5ec269190cce

I would suggest to rework these changes leaving the test env as the default (so removing the setUp) and turning off the upload in the specific test instead than do the opposite, see
https://github.com/DSpace/DSpace/commit/f8ab7def82783da2325c55fbf88d5ec269190cce#diff-38ba2da3dd74a1027593bbd62d684626R79
https://github.com/DSpace/DSpace/commit/f8ab7def82783da2325c55fbf88d5ec269190cce#diff-38ba2da3dd74a1027593bbd62d684626R1713

**Comment by tdonohue** [comment]
@abollini : refactoring the test is fine by me, as long as we still have the same tests.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/UploadAccessConditionDTO.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+import java.util.Date;
+import java.util.UUID;
+
+/**
+ * The UploadAccessCondition it is partial representation of the DSpace ResourcePolicy
+ *
+ * @author Mykhaylo Boychuk (mykhaylo.boychuk at 4science.it)
+ */
+public class UploadAccessConditionDTO  {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/model/UploadAccessConditionDTO.java]: I have to admit, this class is confusing to me. What is an "UploadAccessCondition" and what does it have to do with a ResourcePolicy?  Why do we require a "partial representation" of a ResourcePolicy? Why is this class a DTO, when we tend to create `*Rest` model objects for the REST API layer?

Also, I'm noticing we already have an `org.dspace.app.rest.model.AccessConditionOptionRest` object and an `org.dspace.app.rest.model.ResourcePolicyRest` object.  What is this class doing that couldn't be handled by one of those other existing objects? 

(To clarify here, I'm not against adding this DTO if it's needed. But, I'd like us to better describe *why* it is needed and how it differs from the previously mentioned Rest objects. It's currently not clear to me why we need this object.)

**Review by tdonohue — COMMENTED** [review]
@Micheleboychuk and @abollini : I have to unfortunately admit, I'm confused by the changes in this PR. I don't understand what an "UploadAccessCondition" is and how it differs from a ResourcePolicy or an AccessConditionOption (more on that inline below).  

I think I minimally need better JavaDocs to explain what is going on here (unfortunately it doesn't help that other classes this PR touches have very little JavaDocs as well). It's unclear why we need a different object to take over some (but not all) of the PATCH requests from ResourcePolicy. In other words, it's unclear why `ResourcePolicyAddPatchOperation` and `ResourcePolicyReplacePatchOperation` need to no longer use `ResourcePolicyRest`, while `ResourcePolicyRemovePatchOperation` still uses it.  Is there some better way to align all these PATCH operations since they all deal with ResourcePolicy objects?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/UploadAccessConditionDTO.java (hunk 1 above)
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/model/UploadAccessConditionDTO.java]: yes, we need an additional class to represent the payload used by the rest client to update information in the upload section of a workspaceitem. I will add more information about it in the javadoc, but the main difference is that the `ResourcePolicyRest` object contains a reference to the real Group object but here we only need the uuid (the uri-list approach cannot be used inside the patch). The `AccessConditionOptionRest` instead is used to represent one option to set access in the upload pannel, such as open acces, embargo, leave, etc.

You are of course right about use this new class everywhere in the patch, we will look to fix that

Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/model/UploadAccessConditionDTO.java]: Thanks @abollini for the better JavaDocs here, as it's helped me to better wrap my head around how this is working.  I'll now do a more thorough review.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/step/UploadBitstreamRest.java [hunk 2]
```diff
@@ -16,13 +17,13 @@
 import org.dspace.app.rest.model.BitstreamFormatRest;
 import org.dspace.app.rest.model.CheckSumRest;
 import org.dspace.app.rest.model.MetadataValueRest;
-import org.dspace.app.rest.model.ResourcePolicyRest;
+import org.dspace.app.rest.model.UploadAccessConditionDTO;
 
 public class UploadBitstreamRest extends UploadStatusResponse {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/model/step/UploadBitstreamRest.java]: Could we take this opportunity to add JavaDocs to this class as well, and describe how it works with the new UploadAccessConditionDTO?  I notice the UploadAccessConditionDTO is used here & references this UploadBitstreamRest class in it's JavaDocs, but this class doesn't have it's own description

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/SubmissionService.java [hunk 3]
```diff
@@ -168,8 +169,8 @@ public UploadBitstreamRest buildUploadBitstream(ConfigurationService configurati
 
         for (ResourcePolicy rp : source.getResourcePolicies()) {
             if (ResourcePolicy.TYPE_CUSTOM.equals(rp.getRpType())) {
-                ResourcePolicyRest resourcePolicyRest = converter.toRest(rp, Projection.DEFAULT);
-                data.getAccessConditions().add(resourcePolicyRest);
+                UploadAccessConditionDTO uploadAccessCondition = createAccessConditionFromResourcePolicy(rp);
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/SubmissionService.java]: Similarly, with this change, I'd like to ask if we can add JavaDocs to the `buildUploadBitstream()` method (they are currently missing).  Describing this method would really help clarify the purposes of both `UploadBitstreamRest` (which is is building) and `UploadAccessConditionDTO` (which it uses to pass along access conditions from the Bitstream's resource policies)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/UploadAccessConditionDTO.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+ * include reference to the GroupRest and EPersonRest object, in the upload
+ * section data in a simpler way where such reference are just UUID. Indeed, due
+ * to the fact that the RestModel class are serialized according to the HAL
+ * format and the reference are only exposed in the _links section of the
+ * RestResource it was not possible to use the {@link ResourcePolicyRest} class
+ * directly in the upload section
+ *
+ * @author Mykhaylo Boychuk (mykhaylo.boychuk at 4science.it)
+ */
+public class UploadAccessConditionDTO  {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/model/UploadAccessConditionDTO.java]: Based on the new JavaDocs here & the improvements to other Javadocs, I would say this class might be better named `UploadBitstreamAccessConditionDTO`, as it's really specific to *Bitstreams* and it's used/needed directly by `UploadBitstreamRest` (so it'd be nice to name them similarly)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/factory/impl/ResourcePolicyAddPatchOperation.java [hunk 5]
```diff
@@ -35,7 +36,7 @@
  *
  * @author Luigi Andrea Pascarelli (luigiandrea.pascarelli at 4science.it)
  */
-public class ResourcePolicyAddPatchOperation extends AddPatchOperation<ResourcePolicyRest> {
+public class ResourcePolicyAddPatchOperation extends AddPatchOperation<UploadAccessConditionDTO> {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/factory/impl/ResourcePolicyAddPatchOperation.java]: Based on the changes in this PR, I'm realizing that this PatchOperation class (and **all** of the `ResourcePolicy*PatchOperation` classes) is improperly named.

One of the points of confusion I had with this PR was that I couldn't understand why "generic" Resource Policy PATCH operations would use a Bitstream-specific DTO.   After digging deeper (thanks to @abollini 's JavaDocs), I'm realizing this PATCH class is **not generic** (i.e. it's not handling a patch on the `/api/authz/resourcepolicies` endpoint).  It's also specific to Bitstreams.

Could we use this PR as an opportunity to correct the names of these PatchOperation Classes?  They probably should be named:
* `BitstreamResourcePolicyAddPatchOperation`
* `BitstreamResourcePolicyReplacePatchOperation`
* `BitstreamResourcePolicyRemovePatchOperation`

**Review by tdonohue — COMMENTED** [review]
@abollini : Thanks for the thorough explanations here. The code structure makes a lot more sense after you've described the purposes of each of the classes here.

I have no objections to the code itself here.  However, I do have some requests to complete some additional JavaDocs (see inline below).  I'm also realizing that several of the classes touched here seem **improperly named**.  When I initially looked at this PR, it was very confusing as parts of the code seemed Bitstream-specific, while it seemed to touch classes that *implied* they were not Bitstream-specific.  So, another major point of confusion here is that some classes have names which imply they can perform *non-Bitstream* tasks.  That seems to be false, as I'm realizing all these classes are Bitstream specific.  See my notes inline below on the specific classes which I feel are improperly named.

Overall, now that I understand it, I think the code here now looks good.  Thanks @Micheleboychuk ! However, I'd like to see a few more JavaDocs & some renaming of classes before we merge it.

**Comment by abollini** [comment]
the proposed names make sense to me as well. We will apply the refactoring tomorrow

**Comment by abollini** [comment]
@tdonohue I should have addressed your latest feedbak

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/factory/impl/BitstreamResourcePolicyAddPatchOperation.java [hunk 6]
```diff
@@ -74,17 +76,19 @@ void add(Context context, Request currentRequest, InProgressSubmission source, S
...
 
                         //TODO manage error on select group and eperson
                         Group group = null;
                         if (newAccessCondition.getGroupUUID() != null) {
+                           // UUID uuidGroup = UUID.fromString(newAccessCondition.getGroup().getUuid());
                             group = groupService.find(context, newAccessCondition.getGroupUUID());
                         }
                         EPerson eperson = null;
                         if (newAccessCondition.getEpersonUUID() != null) {
+                           // UUID uuidEPerson = UUID.fromString(newAccessCondition.getEperson().getUuid());
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/factory/impl/BitstreamResourcePolicyAddPatchOperation.java]: There's two commented out lines here that don't seem to be necessary.  Here and 5 lines above this line.

**Review by tdonohue — COMMENTED** [review]
@abollini : This looks good overall, however it looks like you missed on piece of feedback.  I had also recommended renaming the new `UploadAccessConditionDTO` to be `UploadBitstreamAccessConditionDTO`.  See this comment: https://github.com/DSpace/DSpace/pull/2665#discussion_r382275348

One other very minor inline comment below.  Overall, the direction here is good though, just minor cleanup before we can merge.

UPDATE: I've tested this PR and it appears to work as described and seems to fix DS-4418.

**Comment by abollini** [comment]
@tdonohue this should be ready to merge now

**Review by tdonohue — APPROVED** [review]
👍 This looks good to me now.  The requested changes have all been made, and it works as expected/described.  Thanks @Micheleboychuk and @abollini .

Merging as this was flagged for one approval in DSpace 7 WG

**Pull request merged**